### DISCUSSION TITLE
Is the usage of 'bipartite graphs' accurate in this documentation?

### DISCUSSION THREAD

**Discussion by lingvisa** [pyg-team/pytorch_geometric#3667] [General]
I am reading the documentation on graphSage's NeighborSampler:

"Given a GNN with  layers and a specific mini-batch of nodes node_idx for which we want to compute embeddings, this module iteratively samples neighbors and constructs **bipartite graphs** that simulate the actual computation flow of GNNs".

Doesn't 'bipartite graph' mean the graph has at least two types of nodes, i.e. hetero graph? The original graphSage is for homo graph. Does the term 'bipartite graph' has another meaning here in this documentation?

**Comment by rusty1s** [comment]
Sorry for the confusion. Here, a bipartite graph refers to a graph with two node sets of the same type: the set of source nodes and the set of destination nodes. I hope this resolves your concerns :)

**Reply by lingvisa** [reply]
Good to know for the clarification.

**Comment by errhernandez** [comment]
Can I ask for some further clarification here, @rusty1s? When you say "two node sets of the same type" do you mean that the source nodes and destination nodes have features of the same dimension? Is that the difference between bipartite and heterogeneous graphs? That in bipartite graphs nodes are of the same type, i.e. have features of the same dimension, and are distinguished by some being source nodes and others being target nodes, while in heterogeneous graphs there are nodes having different types (different lengths) of feature vectors?

**Reply by rusty1s** [reply]
This comment is more specific to the implementation of `NeighborSampler`. In general, GNN layers in PyG do not have the restriction that the bipartite graph needs to be between the same set of node types. Most GNN layers support something like `SAGEConv((feat_dim_src_node_type, feat_dim_dst_node_type), 64)` such that you can perform bipartite message passing between different feature dimensionalities. Hope this clarifies your doubts.

**Reply by errhernandez** [reply]
Thank you. So, is it correct to say that a bipartite graph is a particular case of an heterogeneous graph with just two types of nodes that differ on their feature dimensionality (feat_dim_src_node_type, feat_dim_dst_node_type in your example above)? From the documentation it seems that in bipartite graphs the flow of information is from src nodes to dst nodes, while in heterogeneous graphs the flow can be in both directions, right? Sorry to ask so many questions!

**Reply by rusty1s** [reply]
Yes, I actually see heterogeneous graphs as just a sequence of bipartite message passing steps between all different edge types.
